### PR TITLE
Feature/12 factor app

### PR DESCRIPTION
**Pull request by cadmiumcat**
### What

Attempt at 12 Factor App documentation

### How to review

Read it and:
- Watch out for any straight up lie I might be telling without realising.
- Watch out for typos

Suggestions welcomed 🙂 

### Who can review

Anyone

### CODE REVIEW

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+### 12. Admin processes
+
+### Where we could improve
+
+## Next steps
+
+
+Further resources
+----------------------------
+- Using [dp-cli to generate a repository](https://github.com/ONSdigital/dp-cli/blob/master/project_generation/COMPLETE_PROJECT_SETUP.md)
```
Comment (nshumoogum) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:69]: Add any resources specified in my previous comments here, yet it is unnecessary to list all 12 links for 12factor.net and just add https://12factor.net and allow someone to navigate that site instead.

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+## Pre requisites
+None.
+
+## Materials
+Familiarise yourself [The Twelve-Factor App Principles](https://12factor.net/).
+
+In this section you can find out how we apply each principle in Digital Publishing:
+
+### 1. Codebase
+We create one GitHub repository per application, for example the [dp-bulletin-api](https://github.com/ONSdigital/dp-bulletin-api) repository contains our codebase for that app.
```
Comment (nshumoogum) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:23]: Worth adding a description of what makes an app a 12 factor app - A twelve-factor app is always tracked in a version control system

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+In Digital Publishing we build, deploy and manage applications which run as services. The twelve-factor methodology helps us follow best practices and build applications which are portable and resilient. This module will introduce you to the Twelve-factor principles and provide you with examples of how we use them in our team. 
+
+Note: The examples refer primarily to Go apps because that's what I know about. Please add more details or alternative examples which might be useful.
+
+## Pre reading
+
+- [About version control](https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control)
+- [What is Continuous Integration (CI) and Continuous Delivery (CD)?](https://www.redhat.com/en/topics/devops/what-is-ci-cd)
+
```
[the remark this answers is not in the document]
Comment (cadmiumcat) [reply training/architecture/12_FACTOR_APP_PRINCIPLES.md:12]: oh I like that 👍

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md (hunk 2 above)
Comment (cadmiumcat) [reply training/architecture/12_FACTOR_APP_PRINCIPLES.md]: Thank you for all the suggestions, they helped me understand the principles better (I think!)

**Review by redhug1 — COMMENTED** [review]
Near the top of the document, i don't think it needs the following:
" because that's what I know about"

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+Note: The examples refer primarily to Go apps because that's what I know about. Please add more details or alternative examples which might be useful.
+
+## Pre reading
+- Introduction and Background to [the twelve-factor app](https://12factor.net/)
+- [About version control](https://git-scm.com/book/en/v2/Getting-Started-About-Version-Control)
+- [What is Continuous Integration (CI) and Continuous Delivery (CD)?](https://www.redhat.com/en/topics/devops/what-is-ci-cd)
+
+
+## Pre requisites
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:14]: Pre requisites is one word

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+See pre-requisites in each section.
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:15]: Same comment as above

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
+12 Factor App Principles
+===========================
+
+In Digital Publishing we build, deploy and manage applications which run as services. The twelve-factor methodology helps us follow best practices and build applications which are portable and resilient. This module will introduce you to the Twelve-factor principles and provide you with examples of how we use them in our team. 
+
+Note: The examples refer primarily to Go apps because that's what I know about. Please add more details or alternative examples which might be useful.
+
+## Pre reading
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:8]: Pre-reading is hyphenated( I think)

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 7]
```diff
@@ -0,0 +1,167 @@
+12 Factor App Principles
+===========================
+
+In Digital Publishing we build, deploy and manage applications which run as services. The twelve-factor methodology helps us follow best practices and build applications which are portable and resilient. This module will introduce you to the Twelve-factor principles and provide you with examples of how we use them in our team.
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:4]: Should it be `in our teams`?

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+The app should be tracked using version control. In DP, We create one GitHub repository per application, for example the [dp-bulletin-api](https://github.com/ONSdigital/dp-bulletin-api) repository contains our codebase for that app.
+Each _deploy_ of the `dp-bulletin-api` will use a version of the codebase. The version is defined by an annotated tag which references a specific commit in the `dp-bulletin-api` repository. The `production` environment will contain a version deployed from the `master`/`main` branch of the repository, which may be different from the `develop` branch.
+
+Find out more:
+- How we [collaborate on Digital Publishing repos](https://github.com/ONSdigital/dp/blob/master/guides/CONTRIBUTING.md).
+- [Semantic Versioning](https://semver.org/)
+
+### 2. Dependencies
+Pre-requisite: [Dependencies](https://12factor.net/dependencies)
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:34]: same as above

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+## Pre requisites
+See pre-requisites in each section.
+
+## Materials
+
+In this section you can find out how we apply each principle in Digital Publishing:
+
+### 1. Codebase
+Pre-requisite: [Codebase](https://12factor.net/codebase)
```
Comment (sandypadmanabhan) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:22]: same comment as above

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+Find out more:
+- How we [collaborate on Digital Publishing repos](https://github.com/ONSdigital/dp/blob/master/guides/CONTRIBUTING.md).
+- [Semantic Versioning](https://semver.org/)
+
+### 2. Dependencies
+Pre-requisite: [Dependencies](https://12factor.net/dependencies)
+
+> Explicitly declare and isolate dependencies
+
+A 12-factor app will need to have ownership of its dependencies by defining the version/release of a library ([what's the difference between and app and a library?](https://stackoverflow.com/questions/1270729/difference-between-library-and-application-code)). This allows for better control of how the application functions instead of always inheriting the latest version of a library which might not be backward compatible.
```
Comment (HeJones) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:38]: should and app be an app?

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+### 6. Processes
+Pre-requisite: [Processes](https://12factor.net/processes)
+
+> Execute the app as one or more stateless processes
+ 
+A 12-factor app is stateless, i.e. it does not save client data from a session to be used in the next session with that client. This means that restarting the app does not result in different outcomes when processing information or requests.
+
+In DP, the data that persist is stored in stateful backing service (database... or static file?)
+
```
Comment (HeJones) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:91]: Should this be a question?

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+- The logs are streamed, centralised and aggregated in [Kibana](https://www.elastic.co/kibana)
+
+### 12. Admin processes
+Pre-requisite: [Admin processes](https://12factor.net/admin-processes)
+
+> Run admin/management tasks as one-off processes
+
+One-off administrative tasks should be one-off processes shipped with application code. Then the process (admin task) should be run remotely, for example by SSHing to the environment.
+
+In DP, sometimes write these scripts in a separate git repo [dp-data-tools](https://github.com/ONSdigital/dp-data-tools) and hence it is not deployable; instead, we use flags to allow us to inject the correct configuration for each environment.
```
Comment (HeJones) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:154]: in DP, we sometimes

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md (hunk 11 above)
Comment (cadmiumcat) [reply training/architecture/12_FACTOR_APP_PRINCIPLES.md:91]: I'm just not sure what I'm saying makes sense 🙈

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+
+### 6. Processes
+Prerequisite: [Processes](https://12factor.net/processes)
+
+> Execute the app as one or more stateless processes
+ 
+A 12-factor app is stateless, i.e. it does not save client data from a session to be used in the next session with that client. This means that restarting the app does not result in different outcomes when processing information or requests.
+
+In DP, the data that persist is stored in stateful backing service (database... or static file?)
```
Comment (gedge) [inline training/architecture/12_FACTOR_APP_PRINCIPLES.md:90]:
```suggestion
In DP, data that should persist is stored in a stateful backing service (typically a database like Neptune or MongoDB, but also a message queue, such as kafka).
```

no need to mention static files: they are deprecated

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md (hunk 11 above)
Comment (gedge) [reply training/architecture/12_FACTOR_APP_PRINCIPLES.md:91]: have added a suggestion

**Review by gedge — CHANGES_REQUESTED** [review]

File: training/architecture/12_FACTOR_APP_PRINCIPLES.md [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+In our Go apps we use the [envconfig package](https://github.com/kelseyhightower/envconfig) to help us implement this principle.
+
+
+### 4. Backing services
+Pre-requisite: [Backing services](https://12factor.net/backing-services)
+
+> Treat backing services as attached resources
+
+Our apps rely on the use of external services, such as databases (e.g. MongoDB).
+A 12-factor app should be able to switch out these external services by modifying the configuration, e.g. with no code changes.
```
[the remark this answers is not in the document]
Comment (cadmiumcat) [reply training/architecture/12_FACTOR_APP_PRINCIPLES.md]: I've changed the others now 🙂

**Review by gedge — DISMISSED** [review]
assuming @nshumoogum is happy...

**Review by gedge — APPROVED** [review]

**Review by nshumoogum — APPROVED** [review]

**Pull request merged**